File: sentimental-credit/credit.py
from cs50 import get_string

# Gets number of card
card = get_string("Number: ")
size = len(card)
sum = 0

# Calculates Luhn's Algorithm
for i in range(size - 2, -1, -2):
    if (int(card[i]) * 2 > 9):
        sum += (int(card[i]) * 2) % 10
        sum += 1
    else:
        sum += int(card[i]) * 2
for i in range(size - 1, -1, -2):
    sum += int(card[i])

# Checks if the card is valid and which card it is
if (sum % 10 == 0):
    if (size == 15 and (card[:2] == "34" or card[:2] == "37")):
        print("AMEX")
    elif (size == 16 and (card[:2] == "51" or card[:2] == "52" or card[:2] == "53" or card[:2] == "54" or card[:2] == "55")):
        print("MASTERCARD")
    elif ((size == 13 or size == 16) and card[0] == "4"):
        print("VISA")
    else:
        print("INVALID")
else:
    print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

void checarCartao(int digito, int primeiros);

int main(void)
{
    int valorUltimo = 0;
    int digitos = 0;
    int valorPenultimo = 0;
    int somaPrimeira = 0;
    int somaSegunda = 0;
    long card = get_long("Number: ");
    long cartaoNovo = card;
    do
    {
        valorUltimo = cartaoNovo % 10;
        cartaoNovo = (cartaoNovo - valorUltimo) / 10;
        if (cartaoNovo == 0)
        {
            digitos++;
        }
        else
        {
            digitos += 2;
        }
        valorPenultimo = cartaoNovo % 10;
        cartaoNovo = (cartaoNovo - valorPenultimo) / 10;

        if (valorPenultimo >= 5)
        {
            int dobro = valorPenultimo * 2;
            valorPenultimo = dobro % 10;
            dobro = (dobro - valorPenultimo) / 10;
            valorPenultimo += dobro;
        }
        else
        {
            valorPenultimo = valorPenultimo * 2;
        }
        somaSegunda += valorUltimo;
        somaPrimeira += valorPenultimo;
    }
    while (cartaoNovo > 0);

    somaPrimeira += somaSegunda;
    card = card / pow(10, digitos - 2);
    int primeirosDigitos = card;

    //Checks the Luhn's algorithm and if it's valid then it calls the function checarCartao and if not it prints "INVALID" in the screen
    if (somaPrimeira % 10 == 0)
    {
        checarCartao(digitos, primeirosDigitos);
    }
    else
    {
        printf("INVALID\n");
    }
}



//Function that receives the number os digits tha card has and the two first digits and prints the name of the card or "INVALID" if not a valid card
void checarCartao(int digito, int primeiros)
{
    if (digito == 15 && (primeiros == 34 || primeiros == 37))
    {
        printf("AMEX\n");
    }
    else if (digito == 16 && (primeiros >= 51 && primeiros <= 55))
    {
        printf("MASTERCARD\n");
    }
    else if ((digito == 13 || digito == 16) && primeiros >= 40 && primeiros <= 49)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}